stream/src/main/java/org/apache/kafka/streaming/processor/TopologyBuilder.java

```diff
@@ -0,0 +1,203 @@
+import org.apache.kafka.streaming.processor.internals.ProcessorTopology;
+import org.apache.kafka.streaming.processor.internals.SinkNode;
+import org.apache.kafka.streaming.processor.internals.SourceNode;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class TopologyBuilder {
+
+    // list of node factories in a topological order
+    private ArrayList<NodeFactory> nodeFactories = new ArrayList<>();
+
+    private Set<String> nodeNames = new HashSet<>();
+    private Set<String> sourceTopicNames = new HashSet<>();
+
+    private interface NodeFactory {
+        ProcessorNode build();
+    }
+
+    private class ProcessorNodeFactory implements NodeFactory {
+        public final String[] parents;
+        private final String name;
+        private final ProcessorDef definition;
+
+        public ProcessorNodeFactory(String name, String[] parents, ProcessorDef definition) {
+            this.name = name;
+            this.parents = parents.clone();
+            this.definition = definition;
+        }
+
+        public ProcessorNode build() {
+            Processor processor = definition.instance();
+            return new ProcessorNode(name, processor);
+        }
+    }
+
+    private class SourceNodeFactory implements NodeFactory {
+        public final String[] topics;
+        private final String name;
+        private Deserializer keyDeserializer;
+        private Deserializer valDeserializer;
+
+        private SourceNodeFactory(String name, String[] topics, Deserializer keyDeserializer, Deserializer valDeserializer) {
+            this.name = name;
+            this.topics = topics.clone();
+            this.keyDeserializer = keyDeserializer;
+            this.valDeserializer = valDeserializer;
+        }
+
+        public ProcessorNode build() {
+            return new SourceNode(name, keyDeserializer, valDeserializer);
+        }
+    }
+
+    private class SinkNodeFactory implements NodeFactory {
+        public final String[] parents;
+        public final String topic;
+        private final String name;
+        private Serializer keySerializer;
+        private Serializer valSerializer;
+
+        private SinkNodeFactory(String name, String[] parents, String topic, Serializer keySerializer, Serializer valSerializer) {
+            this.name = name;
+            this.parents = parents.clone();
+            this.topic = topic;
+            this.keySerializer = keySerializer;
+            this.valSerializer = valSerializer;
+        }
+        public ProcessorNode build() {
+            return new SinkNode(name, topic, keySerializer, valSerializer);
+        }
+    }
+
+    public TopologyBuilder() {}
+
+    public final void addSource(String name, Deserializer keyDeserializer, Deserializer valDeserializer, String... topics) {
```

Comment:
Can `addSource`, `addSink`, and `addProcessor` be changed to return `TopologyBuilder` instance so that the builder's methods can be chained together. If so, then this:

    TopologyBuilder builder = new TopologyBuilder();
    builder.addSource("SOURCE", new StringDeserializer(), new IntegerDeserializer(), "topic-source");
    builder.addProcessor("PROCESS", new MyProcessorDef(), null, "SOURCE");
    builder.addSink("SINK", "topic-sink", new StringSerializer(), new IntegerSerializer(), "PROCESS");
    return builder;

becomes:

    return new TopologyBuilder().addSource("SOURCE", new StringDeserializer(), new IntegerDeserializer(), "topic-source")
                                .addProcessor("PROCESS", new MyProcessorDef(), null, "SOURCE")
                                .addSink("SINK", "topic-sink", new StringSerializer(), new IntegerSerializer(), "PROCESS");

Might not be useful in all situations, but in some cases it works quite beautifully. I'd be happy to submit a pull-request for this.